xcode/Subconscious/Shared/Components/AddressBook/AddressBookModel.swift

```diff
@@ -9,6 +9,57 @@ import os
 import Foundation
 import SwiftUI
 import ObservableStore
+import Combine
+
+typealias Validator<I, O> = (I) -> O?
+
+struct FormField<I : Equatable, O>: Equatable {
+    static func == (lhs: FormField<I, O>, rhs: FormField<I, O>) -> Bool {
+        return lhs.value == rhs.value && lhs.touched == rhs.touched && lhs.isValid == rhs.isValid
+    }
+    
+    var value: I
+    var validate: Validator<I, O>
+    var touched: Bool = false
+    
+    func update(input: I) -> Self {
+        Self(value: input, validate: validate, touched: touched)
+    }
+    
+    /// Only marked as touched when field _loses_ focus
+    func touch(focused: Bool = false) -> Self {
+        Self(value: value, validate: validate, touched: focused ? touched : true)
+    }
+    
+    var validated: O? {
+        get {
+            validate(value)
+        }
+    }
+    var isValid: Bool {
+        get {
+            validated != nil
+        }
+    }
+    var hasError: Bool {
+        get {
+            !isValid && touched
+        }
+    }
+}
+
+struct FollowUserForm: Equatable {
+    var did: FormField<String, Did> = FormField(value: "", validate: Self.validateDid)
+    var petname: FormField<String, Petname> = FormField(value: "", validate: Self.validatePetname)
+    
+    static func validateDid(key: String) -> Did? {
+        Did(key)
+    }
+
+    static func validatePetname(petname: String) -> Petname? {
+        Petname(petname)
+    }
+}
```

Comment:
There's an argument for flattening these fields into the parent `AddressBookModel` instead of nesting.